Libraries/LibCore/Account.cpp

```diff
@@ -0,0 +1,181 @@
+
+    struct passwd* pwd = nullptr;
+    pwd = getpwnam(username_with_terminator.characters());
+    if (!pwd)
+        return String("No such user");
+
+    Account account(pwd, get_gids(pwd->pw_name));
+    endpwent();
+    return account;
+}
+
+Result<Account, String> Account::get(uid_t uid)
+{
+    struct passwd* pwd = nullptr;
+    pwd = getpwuid(uid);
+    if (!pwd)
+        return String("No such user");
+
+    Account account(pwd, get_gids(pwd->pw_name));
+    endpwent();
+    return account;
+}
+
+bool Account::authenticate(const StringView& password) const
+{
+    if (m_passwd == "")
+        return true;
+
+    String password_with_terminator(password);
+
+    // FIXME: Use crypt_r if it can be built in lagom.
+    char* hash = crypt(password_with_terminator.characters(), m_passwd.characters());
+    return hash != nullptr && strcmp(hash, m_passwd.characters()) == 0;
+}
+
+Result<uid_t, int> Account::login() const
+{
+    if (setgroups(m_extra_gids.size(), m_extra_gids.data()) < 0)
+        return errno;
+
+    if (setgid(m_gid) < 0)
+        return errno;
+
+    if (setuid(m_uid) < 0)
+        return errno;
+
+    return m_uid;
+}
+
+void Account::set_passwd(const StringView& passwd)
+{
+    String tmp_passwd(passwd);
+    m_passwd = crypt(tmp_passwd.characters(), get_salt().characters());
+}
+
+void Account::set_passwd_enabled(bool enabled)
+{
+    if (enabled && m_passwd != "" && m_passwd[0] == '!') {
+        m_passwd = m_passwd.substring(1, m_passwd.length() - 1);
+    } else if (!enabled && (m_passwd == "" || m_passwd[0] != '!')) {
+        StringBuilder builder;
+        builder.append('!');
+        builder.append(m_passwd);
+        m_passwd = builder.build();
+    }
+}
+
+void Account::delete_passwd()
+{
+    m_passwd = "";
+}
+
+Account::Account(struct passwd* pwd, Vector<gid_t> extra_gids)
+    : m_username(pwd->pw_name)
+    , m_passwd(pwd->pw_passwd)
+    , m_uid(pwd->pw_uid)
+    , m_gid(pwd->pw_gid)
+    , m_gecos(pwd->pw_gecos)
+    , m_home_directory(pwd->pw_dir)
+    , m_shell(pwd->pw_shell)
+    , m_extra_gids(extra_gids)
+{
+}
+
+void Account::sync()
+{
+    StringBuilder new_passwd_file;
+
+    setpwent();
+
+    struct passwd* p;
+    while ((p = getpwent())) {
+        if (p->pw_uid == m_uid) {
+            new_passwd_file.appendf("%s:%s:%u:%u:%s:%s:%s\n",
+                m_username.characters(),
+                m_passwd.characters(),
+                m_uid, m_gid,
+                m_gecos.characters(),
+                m_home_directory.characters(),
+                m_shell.characters());
+
+        } else {
+            new_passwd_file.appendf("%s:%s:%u:%u:%s:%s:%s\n",
+                p->pw_name, p->pw_passwd, p->pw_uid,
+                p->pw_gid, p->pw_gecos, p->pw_dir,
+                p->pw_shell);
+        }
+    }
+    endpwent();
+
+    auto file = Core::File::open("/etc/passwd", Core::IODevice::OpenMode::WriteOnly);
+    file.value()->write(new_passwd_file.build());
```

Comment:
Reminder that any of these (opening the files, reading entries, writing data) can fail, and you need to nicely propagate the failure to the caller rather than swallowing it and carrying on.

Also, `IODevice::write()` is clearly broken:

https://github.com/SerenityOS/serenity/blob/02b3cb812367266da673edda30c3393fa25e2e52/Libraries/LibCore/IODevice.cpp#L269-L278

Please either fix it, or use some other means of writing to a file.

Reply:
Still relevant: all of these can fail, and `IODevice::write()` doesn't actually write out the whole thing, only a part of it, and doesn't even tell you how much.

Reply:
I haven't gotten to that part yet

Reply:
@awesomekling this is still unsolved, @Petelliott hasn't gotten to that part yet :smile:

Reply:
Ah okay! The GitHub UI is confusing :]

Reply:
It is confusing, yeah!

Still beats sending and reviewing patches over email :slightly_smiling_face:

Reply:
I've changed this to use `stdio` rather than `IODevice`. Sorry for the extremely long delay.

Reply:
Great!

I've got some more suggestions, sorry :(